File: gateway/src/main/java/com/gaskony/scriptide/gateway/presence/Presence.java
package com.gaskony.scriptide.gateway.presence;

import java.util.List;

/**
 * Who else has this file open, and where from.
 *
 * <h2>Why a shared vocabulary</h2>
 *
 * <p>Two very different mechanisms answer the same question — this module's own
 * WebSocket clients, and Ignition's Designer concurrency events — and the person
 * reading the answer does not care which one it came from. They care that
 * somebody else is in the file. So both are normalised into {@link Peer} the
 * moment they arrive, and the difference survives only as {@link Peer#kind()},
 * which the UI shows as a label rather than treating as two features.</p>
 *
 * <h2>What a peer is NOT</h2>
 *
 * <p>It is not a lock. Nothing here refuses a save, and nothing here should ever
 * start to: this module's write path already has optimistic concurrency through
 * {@code If-Match}, which is the mechanism that actually prevents a lost update.
 * Presence exists so two people find out about each other BEFORE the conflict,
 * not so the second one is stopped after it.</p>
 *
 * <p>It is also not proof that somebody is typing. The Designer reports the
 * resources a session has OPEN, which is what its own concurrency banner means
 * too, and a tab left open over lunch counts. Saying "has it open" rather than
 * "is editing" is therefore the accurate wording and is used throughout.</p>
 */
public final class Presence {

    private Presence() { /* vocabulary only */ }

    /** Where a peer is working. */
    public enum Kind {
        /** Another browser connected to this module. */
        IDE,
        /** An Ignition Designer, reported by the platform's own concurrency events. */
        DESIGNER
    }

    /**
     * One other person, in one place, with one set of resources open.
     *
     * @param sessionId opaque and stable for the life of that session; the key
     *                  everything is deduplicated on. Never shown to anyone.
     * @param username  who they are. Blank when the platform did not say — shown
     *                  as "someone", because an empty name in a warning reads as
     *                  a bug in the warning.
     * @param host      the machine they are on, for the case this feature exists
     *                  to serve: the same person, logged in twice, in two places.
     * @param kind      IDE or Designer
     * @param project   the project the session is in
     * @param resources resource paths currently open, in the module/type/name
     *                  form this module uses everywhere else
     * @param since     when the session started, epoch millis, or 0 if unknown
     */
    public record Peer(String sessionId, String username, String host, Kind kind,
                       String project, List<String> resources, long since) {

        public Peer {
            resources = resources == null ? List.of() : List.copyOf(resources);
            username = username == null ? "" : username;
            host = host == null ? "" : host;
            project = project == null ? "" : project;
        }

        /** True when this peer has the given resource path open. */
        public boolean has(String path) {
            return path != null && resources.contains(path);
        }
    }
}

File: web/src/api/presence.ts
/**
 * Who else has this file open, and where they are working from.
 *
 * The gateway keeps the list — see `PresenceRegistry` — and pushes it whenever
 * it changes, from either of two feeds: other browsers connected to this module,
 * and Ignition Designers, which report their open resources to the gateway for
 * their own concurrent-editing banner.
 *
 * ## This is a warning, not a lock
 *
 * Nothing here stops a save. The write path already has optimistic concurrency
 * through `If-Match`, and that is what actually prevents a lost update. Presence
 * exists so two people find out about each other BEFORE the conflict rather than
 * after it, and the wording throughout says "has it open" rather than "is
 * editing" because that is what the Designer reports and what we can honestly
 * claim — a tab left open over lunch counts.
 *
 * ## Reporting is a full list, every time
 *
 * The client sends every path it has open, not a delta. The server replaces that
 * session's set outright, which means a dropped frame self-heals on the next one
 * instead of leaving a name on a file nobody has. It also means the report must
 * be re-sent on RECONNECT: the server's peer entry dies with the socket, so a
 * client that reconnected silently and said nothing would be invisible to
 * everyone else while believing itself announced.
 */
import { sharedTransport } from './lspTransport';

export type PeerKind = 'ide' | 'designer';

/** One other person, in one place. */
export interface Peer {
  /** Blank when the platform did not say — render as "someone", never as "". */
  username: string;
  /** Hostname, or the address where there is no name. */
  host: string;
  kind: PeerKind;
  project: string;
  /** Resource paths, in this module's `<moduleId>/<typeId>/<name>` form. */
  resources: string[];
  /** Session start, epoch millis, or 0 when unknown. */
  since: number;
}

export interface PresenceState {
  peers: Peer[];
  /**
   * Whether per-file Designer presence is working on this gateway.
   *
   * False means the platform's internal event type was not the shape this build
   * expects, so Designer sessions are known at PROJECT level only. The UI says
   * so rather than showing less and looking the same — an indicator that has
   * quietly stopped seeing half its input is worse than one that admits it.
   */
  designerFeed: boolean;
  version: number;
}

export const EMPTY_PRESENCE: PresenceState = { peers: [], designerFeed: false, version: -1 };

type Subscriber = (state: PresenceState) => void;

/**
 * The one presence client for the tab.
 *
 * Shared for the same reason the transport is: two of these would each report
 * their own half of the open tabs, and the last one to send would win.
 */
class PresenceClient {
  private state: PresenceState = EMPTY_PRESENCE;

  private readonly subscribers = new Set<Subscriber>();

  /** The last thing reported, replayed on reconnect. */
  private reported: { project: string; open: string[] } | null = null;

  private started = false;

  private start() {
    if (this.started) return;
    this.started = true;
    const transport = sharedTransport();
    transport.on('presence', (msg) => {
      const payload = msg as Partial<PresenceState> | undefined;
      if (!payload || !Array.isArray(payload.peers)) return;
      this.state = {
        peers: payload.peers as Peer[],
        designerFeed: payload.designerFeed === true,
        version: typeof payload.version === 'number' ? payload.version : 0,
      };
      for (const subscriber of [...this.subscribers]) subscriber(this.state);
    });
    // The server's entry for us died with the old socket, so a reconnect has to
    // re-announce or we are invisible to everyone while believing otherwise.
    transport.onOpen(() => {
      if (this.reported) {
        transport.send('presence', this.reported);
      }
    });
  }

  subscribe(subscriber: Subscriber): () => void {
    this.start();
    this.subscribers.add(subscriber);
    subscriber(this.state);
    return () => {
      this.subscribers.delete(subscriber);
    };
  }

  /** Tell the gateway what this browser has open. Idempotent. */
  report(project: string, open: string[]) {
    this.start();
    const next = { project, open: [...open].sort() };
    if (
      this.reported
      && this.reported.project === next.project
      && this.reported.open.length === next.open.length
      && this.reported.open.every((path, index) => path === next.open[index])
    ) {
      // Nothing changed. Re-sending on every render would put a frame on the
      // wire per keystroke, and each one costs every other client a broadcast.
      return;
    }
    this.reported = next;
    sharedTransport().send('presence', next);
  }
}

let client: PresenceClient | null = null;

export function presenceClient(): PresenceClient {
  if (!client) client = new PresenceClient();
  return client;
}

/** Everyone except you who has this exact path open, in this project. */
export function peersOn(state: PresenceState, project: string, path: string): Peer[] {
  return state.peers.filter(
    (peer) => peer.project === project && peer.resources.includes(path)
  );
}

/** A peer's name as it is shown. Never blank. */
export function peerName(peer: Peer): string {
  return peer.username.trim() || 'someone';
}

/** Where a peer is, as one short phrase: `sam in the Designer on sams-laptop`. */
export function describePeer(peer: Peer): string {
  const where = peer.kind === 'designer' ? 'in the Designer' : 'in the IDE';
  return peer.host ? `${peerName(peer)} ${where} on ${peer.host}` : `${peerName(peer)} ${where}`;
}

/**
 * The sentence shown above a file somebody else also has open.
 *
 * Names everyone rather than counting them: "2 others" makes you go looking,
 * and the whole point is that you already know who to talk to.
 */
export function describeCrowd(peers: Peer[]): string {
  if (peers.length === 0) return '';
  const described = peers.map(describePeer);
  if (described.length === 1) return `${described[0]} also has this open.`;
  const last = described[described.length - 1];
  return `${described.slice(0, -1).join(', ')} and ${last} also have this open.`;
}
